File: shortDrama_framework/tools/paddle-ocr-worker/exporters.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from frame_extract import VideoInfo
from subtitle_merge import DialogueSegment


def export_outputs(
    *,
    output_dir: Path,
    episode_id: str,
    source_video: str,
    video_info: VideoInfo,
    config: dict[str, Any],
    stats: dict[str, int],
    segments: list[DialogueSegment],
) -> dict[str, Path]:
    output_dir.mkdir(parents=True, exist_ok=True)
    dialogue_path = output_dir / "dialogue_segments.json"
    srt_path = output_dir / "ocr.srt"
    meta_path = output_dir / "meta.json"

    dialogue_payload = {
        "episode_id": episode_id,
        "source_video": source_video,
        "segments": [
            {
                "line_id": f"{episode_id}_line_{index + 1:04d}",
                "start": segment.start,
                "end": segment.end,
                "text": segment.text,
                "source": "ocr",
                "role_id": None,
                "role_name": None,
                "emotion": None,
                "need_review": False,
            }
            for index, segment in enumerate(segments)
        ],
    }

    meta_payload = {
        "episode_id": episode_id,
        "source_video": source_video,
        "video_info": {
            "width": video_info.width,
            "height": video_info.height,
            "fps": video_info.fps,
            "duration": video_info.duration,
            "total_frames": video_info.total_frames,
        },
        "ocr_config": config,
        "stats": stats,
    }

    dialogue_path.write_text(json.dumps(dialogue_payload, ensure_ascii=False, indent=2) + "\n", "utf-8")
    srt_path.write_text(to_srt(segments), "utf-8")
    meta_path.write_text(json.dumps(meta_payload, ensure_ascii=False, indent=2) + "\n", "utf-8")

    return {
        "dialogue": dialogue_path,
        "srt": srt_path,
        "meta": meta_path,
    }


def to_srt(segments: list[DialogueSegment]) -> str:
    blocks = []
    for index, segment in enumerate(segments, start=1):
        blocks.append(
            f"{index}\n{format_srt_time(segment.start)} --> {format_srt_time(segment.end)}\n{segment.text}\n"
        )
    return "\n".join(blocks)


def format_srt_time(seconds: float) -> str:
    total_ms = max(0, round(seconds * 1000))
    ms = total_ms % 1000
    total_seconds = total_ms // 1000
    sec = total_seconds % 60
    total_minutes = total_seconds // 60
    minute = total_minutes % 60
    hour = total_minutes // 60
    return f"{hour:02d}:{minute:02d}:{sec:02d},{ms:03d}"

File: shortDrama_framework/apps/server/src/services/videoOcr/exporters.ts
import fs from "node:fs/promises";
import path from "node:path";
import type { DialogueSegment, ExportedDialogueSegment, VideoInfo, VideoOcrConfig } from "./types.js";

export async function exportOcrOutputs(input: {
  outputDir: string;
  episodeId: string;
  sourceVideo: string;
  segments: DialogueSegment[];
  videoInfo: VideoInfo;
  config: VideoOcrConfig;
  stats: {
    sampledFrames: number;
    rawOcrItems: number;
    mergedSegments: number;
  };
}) {
  await fs.mkdir(input.outputDir, { recursive: true });

  const exportedSegments = input.segments.map<ExportedDialogueSegment>((segment, index) => ({
    line_id: `${input.episodeId}_line_${String(index + 1).padStart(4, "0")}`,
    start: segment.start,
    end: segment.end,
    text: segment.text,
    source: "ocr",
    role_id: null,
    role_name: null,
    emotion: null,
    need_review: false
  }));

  const dialogueJsonPath = path.join(input.outputDir, "dialogue_segments.json");
  const srtPath = path.join(input.outputDir, "ocr.srt");
  const metaPath = path.join(input.outputDir, "meta.json");

  await fs.writeFile(
    dialogueJsonPath,
    `${JSON.stringify(
      {
        episode_id: input.episodeId,
        source_video: input.sourceVideo,
        segments: exportedSegments
      },
      null,
      2
    )}\n`,
    "utf8"
  );
  await fs.writeFile(srtPath, toSrt(input.segments), "utf8");
  await fs.writeFile(
    metaPath,
    `${JSON.stringify(
      {
        episode_id: input.episodeId,
        source_video: input.sourceVideo,
        video_info: input.videoInfo,
        ocr_config: {
          ocr_engine: input.config.ocrEngine,
          sample_fps: input.config.sampleFps,
          crop_left_ratio: input.config.cropLeftRatio,
          crop_right_ratio: input.config.cropRightRatio,
          crop_top_ratio: input.config.cropTopRatio,
          crop_bottom_ratio: input.config.cropBottomRatio,
          preprocess_mode: input.config.preprocessMode,
          preprocess_scale: input.config.preprocessScale,
          similarity_threshold: input.config.similarityThreshold,
          min_text_length: input.config.minTextLength,
          max_text_length: input.config.maxTextLength,
          min_han_chars: input.config.minHanChars,
          min_ocr_confidence: input.config.minOcrConfidence,
          max_merge_gap_sec: input.config.maxMergeGapSec,
          ocr_command: input.config.ocrCommand ?? null,
          ocr_languages: input.config.ocrLanguages,
          tesseract_psm: input.config.tesseractPsm,
          tesseract_lang_path: input.config.tesseractLangPath ?? null,
          tesseract_cache_path: input.config.tesseractCachePath ?? null
        },
        stats: input.stats
      },
      null,
      2
    )}\n`,
    "utf8"
  );

  return {
    dialogueJsonPath,
    srtPath,
    metaPath
  };
}

function toSrt(segments: DialogueSegment[]) {
  return segments
    .map(
      (segment, index) =>
        `${index + 1}\n${formatSrtTime(segment.start)} --> ${formatSrtTime(segment.end)}\n${segment.text}\n`
    )
    .join("\n");
}

function formatSrtTime(seconds: number) {
  const totalMs = Math.max(0, Math.round(seconds * 1000));
  const ms = totalMs % 1000;
  const totalSeconds = Math.floor(totalMs / 1000);
  const sec = totalSeconds % 60;
  const totalMinutes = Math.floor(totalSeconds / 60);
  const min = totalMinutes % 60;
  const hour = Math.floor(totalMinutes / 60);

  return `${pad(hour)}:${pad(min)}:${pad(sec)},${String(ms).padStart(3, "0")}`;
}

function pad(value: number) {
  return String(value).padStart(2, "0");
}
